aptos-move/block-executor/src/view.rs

```diff
@@ -86,43 +90,126 @@ pub(crate) enum ReadResult {
+        kind: &ReadKind,
+    ) -> Result<Self, PanicError> {
+        // We set an arbitrary version, as below (from_data_read) internally ignores it.
+        match DataRead::from_value_with_layout(Err(StorageVersion), value).convert_to(kind) {
+            Some(data_read) => Ok(Self::from_data_read(data_read)),
+            None => Err(code_invariant_error(format!(
+                "Could not convert value to kind {:?} ReadResult",
+                kind
+            ))),
+        }
+    }
+
+    pub(crate) fn from_data_read<V: TransactionWrite>(data: DataRead<V>) -> Self {
+        match data {
+            DataRead::Versioned(_, v, layout) => {
+                GroupReadResult::Value(v.extract_raw_bytes(), layout)
             },
-            (ValueWithLayout::RawFromStorage(_), ReadKind::Value) => None,
-            (ValueWithLayout::Exchanged(v, _), ReadKind::Metadata)
-            | (ValueWithLayout::RawFromStorage(v), ReadKind::Metadata) => {
-                Some(ReadResult::Metadata(v.as_state_value_metadata()))
+            DataRead::Resolved(_) => {
+                // Resolved is only available in data MVHashMap for legacy AggregatorV1.
+                unreachable!("Resolved is not a possible group read result");
             },
-            (ValueWithLayout::Exchanged(v, _), ReadKind::Exists)
-            | (ValueWithLayout::RawFromStorage(v), ReadKind::Exists) => {
-                Some(ReadResult::Exists(!v.is_deletion()))
+            DataRead::MetadataAndResourceSize(_, _) | DataRead::Metadata(_) => {
+                // Metadata for the group does not go through the group MVHashMap and is handled
+                // separately in view.rs (it's stored in the data MVHashMap).
+                unreachable!("Metadata may not be queried for resource group members");
             },
+            DataRead::ResourceSize(maybe_size) => GroupReadResult::ResourceSize(maybe_size),
+            DataRead::Exists(exists) => GroupReadResult::Exists(exists),
         }
     }
 
-    pub fn into_value(self) -> Option<StateValue> {
-        if let ReadResult::Value(v, _layout) = self {
-            v
-        } else {
-            unreachable!("Read result must be Value kind")
+    pub fn into_value(self) -> Option<Bytes> {
+        match self {
+            GroupReadResult::Value(maybe_bytes, _) => maybe_bytes,
+            GroupReadResult::GroupSize(_) => {
+                unreachable!("Expected group value, found size")
+            },
+            GroupReadResult::Uninitialized => {
+                unreachable!("Expected group value, found uninitialized")
+            },
+            GroupReadResult::ResourceSize(_) => {
+                unreachable!("Expected group value, found resource size")
+            },
+            GroupReadResult::Exists(_) => {
+                unreachable!("Expected group value, found exists")
+            },
+        }
+    }
+
+    pub fn into_group_size(self) -> ResourceGroupSize {
```

Comment:
nit: `expect_group_size` to indicate it can panic? Does this have to be public? Same for above